mqtt-client/src/main/java/org/fusesource/mqtt/client/CallbackConnection.java

```diff
@@ -145,6 +147,28 @@ public void connect(final Callback<Void> cb) {
     }
 
     void reconnect() {
+        if( isReconnecting ) {
```

Comment:
I'm not sure this will properly guard the synchronization of the code. It might be better to use mutex or lock pattern to do it.
On the other hand the whole architecture of this lib async and I see there's also a version (a bit different) that is doing it that way. I wonder if it'd be better to try to use that everywhere instead of this?

Reply:
I'm not sure if a mutex or lock would fix the issue. The issue that i experienced was that the client tried to reconnect thousands of times during the offline period of the broker and when the broker was online again it send those reconnect attempts to the broker. I think that if we use a lock or a mutex here we'll still create a lot of the transport objects although maybe slower than before. The thing that i tried was to implement a way that only one reconnect is running and a second reconnect can only be attempted after the first is done.

Reply:
I get your point. What I was trying to say is that this code is not thread safe. You should consider using AtomicBoolean in a pattern like this

http://kirkwylie.blogspot.com/2010/08/java-initialization-barrier-pattern.html